internal/jobobject/limits.go

```diff
@@ -0,0 +1,324 @@
+		}
+	default:
+		return 0, errors.New("invalid job object cpu rate control type")
+	}
+
+	return info.Value, nil
+}
+
+// SetCPUAffinity sets the processor affinity for the job object represented by
+// `jobHandle`. The affinity is passed in as a bitmask.
+func (job *JobObject) SetCPUAffinity(affinityBitMask uint64) error {
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	if job.handle == 0 {
+		return ErrAlreadyClosed
+	}
+
+	info, err := getExtendedInformation(job.handle)
+	if err != nil {
+		return err
+	}
+	info.BasicLimitInformation.LimitFlags |= uint32(windows.JOB_OBJECT_LIMIT_AFFINITY)
+	info.BasicLimitInformation.Affinity = uintptr(affinityBitMask)
+	return setExtendedInformation(job.handle, info)
+}
+
+// GetCPUAffinity gets the processor affinity for the job object represented by
+// `jobHandle`. The returned affinity is a bitmask.
+func (job *JobObject) GetCPUAffinity() (uint64, error) {
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	if job.handle == 0 {
+		return 0, ErrAlreadyClosed
+	}
+
+	info, err := getExtendedInformation(job.handle)
+	if err != nil {
+		return 0, err
+	}
+	return uint64(info.BasicLimitInformation.Affinity), nil
+}
+
+// SetIOLimit sets the IO limits specified on the job object represented by `jobHandle`.
+func (job *JobObject) SetIOLimit(maxBandwidth, maxIOPS int64) error {
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	if job.handle == 0 {
+		return ErrAlreadyClosed
+	}
+
+	ioInfo, err := getIOLimit(job.handle)
+	if err != nil {
+		return err
+	}
+	ioInfo.ControlFlags |= winapi.JOB_OBJECT_IO_RATE_CONTROL_ENABLE
+	if maxBandwidth != 0 {
+		ioInfo.MaxBandwidth = maxBandwidth
+	}
+	if maxIOPS != 0 {
+		ioInfo.MaxIops = maxIOPS
+	}
+	if _, err := winapi.SetIoRateControlInformationJobObject(job.handle, ioInfo); err != nil {
+		return errors.Wrap(err, "failed to set IO limit info on job object")
+	}
+	return nil
+}
+
+// GetIOMaxBandwidthLimit gets the max bandwidth for the job object represented by `jobHandle`.
+func (job *JobObject) GetIOMaxBandwidthLimit() (int64, error) {
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	if job.handle == 0 {
+		return 0, ErrAlreadyClosed
+	}
+
+	info, err := getIOLimit(job.handle)
+	if err != nil {
+		return 0, err
+	}
+	return info.MaxBandwidth, nil
+}
+
+// GetIOMaxIopsLimit gets the max iops for the job object represented by `jobHandle`.
+func (job *JobObject) GetIOMaxIopsLimit() (int64, error) {
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	if job.handle == 0 {
+		return 0, ErrAlreadyClosed
+	}
+
+	info, err := getIOLimit(job.handle)
+	if err != nil {
+		return 0, err
+	}
+	return info.MaxIops, nil
+}
+
+// Helper function for getting a job object's extended information.
+// Calling function must be holding job object's handle lock when calling this.
+func getExtendedInformation(jobHandle windows.Handle) (*windows.JOBOBJECT_EXTENDED_LIMIT_INFORMATION, error) {
```

Comment:
Even if these aren't exported, can this just take a `*JobObject` receiver so we don't have to pass in the handle? Feels weird these are the only ones we'd swap things up

Reply:
Yeah we can do that, though we do require the calling functions here to hold the job's handle lock, so it should be safe. I'll update though.